File: problems/2135_count-words-obtained-after-adding-a-letter/solution.java
import java.util.*;

class Solution {

    public int wordCount(String[] startWords, String[] targetWords) {
        HashSet<Integer> starts = new HashSet<>();
        for (String w : startWords) {
            starts.add(mask(w));
        }
        int count = 0;
        for (String t : targetWords) {
            int m = mask(t);
            for (int bit = 0; bit < 26; bit++) {
                if ((m & (1 << bit)) != 0 && starts.contains(m ^ (1 << bit))) {
                    count++;
                    break;
                }
            }
        }
        return count;
    }

    private int mask(String w) {
        int m = 0;
        for (int i = 0; i < w.length(); i++) {
            m |= 1 << (w.charAt(i) - 'a');
        }
        return m;
    }
}

File: problems/2135_count-words-obtained-after-adding-a-letter/solution.cpp
class Solution {
  public:
    int wordCount(vector<string> &startWords, vector<string> &targetWords) {
        auto maskOf = [](const string &w) {
            int m = 0;
            for (char c : w) {
                m |= 1 << (c - 'a');
            }
            return m;
        };

        unordered_set<int> starts;
        for (const string &w : startWords) {
            starts.insert(maskOf(w));
        }
        int count = 0;
        for (const string &t : targetWords) {
            int m = maskOf(t);
            for (int bit = 0; bit < 26; bit++) {
                if ((m & (1 << bit)) && starts.count(m ^ (1 << bit))) {
                    count++;
                    break;
                }
            }
        }
        return count;
    }
};
